luci-mod-network: add stricter wireless interface name validation

crass opened the pull request:
@hnyman 
Linux wireless interface names have the following restrictions:

 * It must not be an empty string
 * It must not be '.' or '..'
 * It must not contain any /, : or space character ( , \t, \n, ...)
 * It must be less than 16 chars
 * It likely must not contain any % either

Fixes: 8673aef8db luci-mod-network: remove uciname validation from wireless interface

This is discussed in #6101 and @jow- dug up the [restrictions on the interface name](https://elixir.bootlin.com/linux/latest/source/net/core/dev.c#L995). He suggests that `%` probably can't be used either, but provides no documentation of this. I've opted to disallow `%` to err on the side of caution. The validation is done as a datatype to make it reuseable. This has not been tested as I'm currently not setup to test this. I think its important to have the validation be stricter because allowing names which the kernel will not accept is a bad user experience. Everything looks like it succeeded when changing the interface name, but the interface will not come up and the user is left wondering why, which is not obvious.

jow- commented:
Thank you for your effort @crass - unfortunately the change is incorrect due to the following reasons:

  - The wireless page uses JavaScript datatype validators in https://github.com/openwrt/luci/blob/master/modules/luci-base/htdocs/luci-static/resources/validation.js while you patched the legacy Lua validator functions
  - The `!=` operator is invalid syntax in Lua, it must be `~=`
  - The `%` character in Lua match patterns is special and must be escaped using `%%`
  - The "all space characters" character class must be written as `%s` instead of `\s`
  - The match pattern is not anchored to the string end (`...$`) so it would match interface names starting with one or more legal chars but followed by an illegal one
  
My assumption about disallowed `%` came from the fact that some drivers/places in the kernel use interface name patterns such as `eth%d` upon netdev registration, leaving it to the kernel to pick the next free ordinal to form a device name. I did some tests with `ip link` now and indeed it seems that it is legal to request an ifname containing `%d` where the kernel will replace `%d` with the next free number. I don't think that this behaviour is useful though as it would result in a (non-deterministic) device name different from the one requested.

```
$ sudo ip link add name br% type bridge
RTNETLINK answers: Invalid argument
$ sudo ip link add name br%d type bridge
$ ip link
...
24: br0: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/ether a6:b0:b4:6d:b8:54 brd ff:ff:ff:ff:ff:ff
```

Note how a sole `%` is invalid though, it *must* be followed by `d`, any other combination seems to be disallowed:

```
$ sudo ip link add name br%a type bridge
RTNETLINK answers: Invalid argument
$ sudo ip link add name br%i type bridge
RTNETLINK answers: Invalid argument
$ sudo ip link add name br%s type bridge
RTNETLINK answers: Invalid argument
$ sudo ip link add name br% type bridge
RTNETLINK answers: Invalid argument
$ sudo ip link add name br%% type bridge
RTNETLINK answers: Invalid argument
$
```

crass commented:
> Thank you for your effort @crass - unfortunately the change is incorrect due to the following reasons:

And thank you for looking at this. My unfamiliarity with both the source and Lua are obvious :)

>     * The wireless page uses JavaScript datatype validators in https://github.com/openwrt/luci/blob/master/modules/luci-base/htdocs/luci-static/resources/validation.js while you patched the legacy Lua validator functions

Yep, did a simple search and that seemed at a glance to be where validators were defined.  Since its legacy, I should drop that change and make a change to `validation.js`, correct?

>     * The `!=` operator is invalid syntax in Lua, it must be `~=`
> 
>     * The `%` character in Lua match patterns is special and must be escaped using `%%`
> 
>     * The "all space characters" character class must be written as `%s` instead of `\s`

The three points should not apply is I switch to using a javascript validator, correct?

>     * The match pattern is not anchored to the string end (`...$`) so it would match interface names starting with one or more legal chars but followed by an illegal one

Indeed.

> 
> My assumption about disallowed `%` came from the fact that some drivers/places in the kernel use interface name patterns such as `eth%d` upon netdev registration, leaving it to the kernel to pick the next free ordinal to form a device name. I did some tests with `ip link` now and indeed it seems that it is legal to request an ifname containing `%d` where the kernel will replace `%d` with the next free number. I don't think that this behaviour is useful though as it would result in a (non-deterministic) device name different from the one requested.
> 
> ```
> $ sudo ip link add name br% type bridge
> RTNETLINK answers: Invalid argument
> $ sudo ip link add name br%d type bridge
> $ ip link
> ...
> 24: br0: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000
>     link/ether a6:b0:b4:6d:b8:54 brd ff:ff:ff:ff:ff:ff
> ```
> 
> Note how a sole `%` is invalid though, it _must_ be followed by `d`, any other combination seems to be disallowed:
> 
> ```
> $ sudo ip link add name br%a type bridge
> RTNETLINK answers: Invalid argument
> $ sudo ip link add name br%i type bridge
> RTNETLINK answers: Invalid argument
> $ sudo ip link add name br%s type bridge
> RTNETLINK answers: Invalid argument
> $ sudo ip link add name br% type bridge
> RTNETLINK answers: Invalid argument
> $ sudo ip link add name br%% type bridge
> RTNETLINK answers: Invalid argument
> $
> ```

Ok, so then disallowing `%` seems justified.

crass commented:
I've changed to using the newer javascript validation framework. Also, I've added a description of the constrains in the placeholder text for the field.

jow- commented:
Merged via 6ad6a241c32ebd4d62f6021a2e3cbe4e1bc192c2 - thanks!
